chalk-integration/src/lowering.rs

```diff
@@ -47,12 +50,12 @@ struct Env<'k> {
 /// ```
 #[derive(Debug, PartialEq, Eq)]
 struct AssociatedTyLookup {
-    id: chalk_ir::TypeId<ChalkIr>,
+    id: chalk_ir::AssocTypeId<ChalkIr>,
     addl_parameter_kinds: Vec<chalk_ir::ParameterKind<()>>,
 }
 
-enum NameLookup {
-    Type(chalk_ir::TypeKindId<ChalkIr>),
+enum TypeLookup {
+    Struct(chalk_ir::StructId<ChalkIr>),
```

Comment:
So, something feels weird to me about having the name of this enum be `TypeLookup`, but only be able to be a struct or parameter. Maybe I need to look more at what this code does and is used for, but it seems...weird.

Reply:
The name choice was intentional: structs and parameters are the only types in the system. Perhaps the thing that is *actually* odd is using the term `StructId` when "structs" in Chalk are really meant to represent any user-defined type (i.e., structs, enums, and unions). I was actually considered renaming `StructId` to `UserTypeId` or something else like that.

In particular, traits are not types -- i.e., you can't say that value's type is `Iterator`. You *can* say that it is `dyn Iterator`, but in that case the role of the trait is not to *be* a type but to *bound* a type. In other words, that is short for `exists<T> { T: Iterator }` (this is particularly explicit in chalk's representation), and `T` is the type.

Reply:
Ah, I see. In that case, I'm fine with these changes. Changing `Struct` to `UserTypeId` sounds less nice. At that point, just go to `TypeId`.